presto-hive/src/main/java/parquet/io/ColumnIOConverter.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package parquet.io;
+
+import com.facebook.presto.hive.parquet.Field;
+import com.facebook.presto.hive.parquet.GroupField;
+import com.facebook.presto.hive.parquet.PrimitiveField;
+import com.facebook.presto.hive.parquet.RichColumnDescriptor;
+import com.facebook.presto.spi.type.MapType;
+import com.facebook.presto.spi.type.NamedTypeSignature;
+import com.facebook.presto.spi.type.Type;
+import com.facebook.presto.spi.type.TypeSignatureParameter;
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+import java.util.Locale;
+import java.util.Optional;
+
+import static com.facebook.presto.hive.parquet.ParquetTypeUtils.getArrayElementColumn;
+import static com.facebook.presto.hive.parquet.ParquetTypeUtils.getMapKeyValueColumn;
+import static com.facebook.presto.spi.type.StandardTypes.ARRAY;
+import static com.facebook.presto.spi.type.StandardTypes.MAP;
+import static com.facebook.presto.spi.type.StandardTypes.ROW;
+
+/**
+ * Placed in parquet.io package to have access to ColumnIO getRepetitionLevel() and getDefinitionLevel() methods.
+ */
+public class ColumnIOConverter
+{
+    private ColumnIOConverter()
+    {
+    }
+
+    public static Optional<Field> constructField(Type type, ColumnIO columnIO)
+    {
+        if (columnIO == null) {
+            return Optional.empty();
+        }
+        boolean required = columnIO.getType().getRepetition() != parquet.schema.Type.Repetition.OPTIONAL;
+        int repetitionLevel = columnIO.getRepetitionLevel();
+        int definitionLevel = columnIO.getDefinitionLevel();
+        if (ROW.equals(type.getTypeSignature().getBase())) {
+            GroupColumnIO groupColumnIO = (GroupColumnIO) columnIO;
+            List<Type> parameters = type.getTypeParameters();
+            ImmutableList.Builder<Optional<Field>> fileldsBuilder = ImmutableList.builder();
+            List<TypeSignatureParameter> fields = type.getTypeSignature().getParameters();
+            boolean structHasParameters = false;
+            for (int i = 0; i < fields.size(); i++) {
+                NamedTypeSignature namedTypeSignature = fields.get(i).getNamedTypeSignature();
+                String name = namedTypeSignature.getName().get().toLowerCase(Locale.ENGLISH);
```

Comment:
`namedTypeSignature.getName()` is Optional so get call may fail.

Reply:
Struct fields are always named (unlike, for instance, array and map inner elements). This is why we can safely call `get()` here.